### ISSUE TITLE
Ignore version bump for non-conventional commits

### ISSUE THREAD

**Issue by ivandov** [release-it/conventional-changelog#22]
In most conventions supported with `conventional-changelog`, there are commit types that should not trigger a version bump. 

For example, when using the `angular` or `conventionalcommits` `preset`, version bumps should only be applicable if using commits that begin with `fix`, `feat`, or other types that include a `!` to signify a `BREAKING CHANGE`. 

For commit messages that either do not conform to the types for the specified preset or are not one of the types listed above that signify a version change, `release-it` and this plugin should ignore the creation of a new semver release. 

Is this achievable by this plug-in, since it overrides the standard version bump logic that happens in `release-it`?

**Comment by webpro** [comment]
The options provided to this plugin are sent to its underlying dependency: https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-recommended-bump#readme

However, at first glance I don't see an option for what you are trying to achieve. But maybe it's standard behavior?

**Comment by ivandov** [comment]
Right, my experience here was from using `semantic-release` and how it would ignore version bumps unless commit messages actually matched the criteria outlined in the convention specified.

**Comment by vonovak** [comment]
Hello @webpro, I'm wondering, has this been fixed, or do you plan to fix it? 
I'd love to run release-it in CI but if it makes releases for commits that should not trigger one, it's not possible.
Thank you 🙂

**Comment by webpro** [comment]
Looking into it again, I'm seeing basically all presets for recommended bump work the same. Here's the angular preset: https://github.com/conventional-changelog/conventional-changelog/blob/master/packages/conventional-changelog-angular/conventional-recommended-bump.js (this is the `whatBump` function).

It always returns value `0`, `1` or `2` for `level`, never `null` or `undefined`, always resulting in a value of this array here: https://github.com/conventional-changelog/conventional-changelog/blob/master/packages/conventional-recommended-bump/index.js#L10

However, I just released a small [update of release-it](https://github.com/release-it/release-it/releases/tag/15.0.0-esm.6) (warning: it's only in the upcoming release-it v15), which together with a custom `whatBump` option could do what you want. I've added a spec in this repo (warning: only in v5): https://github.com/release-it/conventional-changelog/commit/124351c3d05a98263b168bab5cca033942ec3edb

So, with release-it v15 and this plugin v5 this would allow something like this in `.release-it.js`:

```js
module.exports = {
  plugins: {
    '@release-it/conventional-changelog': {
      whatBump: commits => ({ level: null, reason: 'Parsed commits do not warrant a version bump.' })
    }
  }
};
```

(You could use this already using the `esm` npm dist tags. Planning to release the majors shortly, though.)

**Comment by ivandov** [comment]
@webpro How does your example code snippet leverage `whatBump` to properly decide when to send `null` for the `level`? 

Are you suggesting that the function would be extended with additional conditional logic to properly determine when to set the `null` value?

**Comment by webpro** [comment]
> @webpro How does your example code snippet leverage `whatBump` to properly decide when to send `null` for the `level`?
> 
> Are you suggesting that the function would be extended with additional conditional logic to properly determine when to set the `null` value?

My code example is only to indicate where the function fits in the release-it/plugin config.

So yes, additional logic is required, much like the original (linked to above).

**Comment by jamesvillarrubia** [comment]
Following up on this.  I have created a whatBump that seems to align more with the intent of the angular preset and the semantic-release version.

Here's the function:
```javascript
whatBump: (commits,options)=>{
        // return null
        // (commits) => {
          let defaults = {
            build: 'ignore',
            ci: 'ignore',
            docs: 'ignore',
            feat: 'minor',
            fix: 'patch',
            perf: 'patch',
            refactor: 'ignore',
            test: 'ignore'
          }
          let types = (options?.preset?.types || [])
          .reduce((a, v) => {
            return { ...a, [v.type]: v.release}
          }, {}) 

          types = Object.assign({},defaults,types)
          let breakings = 0
          let features = 0
          let levelSet = ['major','minor','patch','ignore']
          let level = Math.min.apply(Math, commits.map(commit => {
            let level = levelSet.indexOf(types[commit.type])
            level = level<0?3:level
            if (commit.notes.length > 0) {
              breakings += commit.notes.length
            }
            if(commit.type === 'feat'){
              features += 1;
            }
            return level
          }))
      
          return {
            level: level,
            reason: breakings === 1
              ? `There is ${breakings} BREAKING CHANGE and ${features} features`
              : `There are ${breakings} BREAKING CHANGES and ${features} features`
          }
}
```

You can then combine this with:
```
"preset": {
        "name": "conventionalcommits",
        "types": [
          {
            "type": "refactor",
            "release": "patch"
          },
          {
            "type": "style",
            "release": "patch"
          },
          {
            "type": "perf",
            "release": "patch"
          },
          {
            "type": "chore",
            "release": "patch"
          },
          {
            "type": "ci",
            "release": "patch"
          }
        ]
      }
 ```
 
 This gives you the ability to override the specific types with one of  `['major', 'minor', 'patch', 'ignore']`.  You can also set your own types.

**Comment by jamesvillarrubia** [comment]
I'm happy to make a PR if it's deemed appropriate.  Possibly to the `angular` plugin or the `conventionalcommit`. @webpro

**Comment by webpro** [comment]
That would be nice, @jamesvillarrubia! Are you still up for this? Would be great to have options to ignore certain types of commits.

**Comment by mogelbrod** [comment]
> Following up on this. I have created a whatBump that seems to align more with the intent of the angular preset and the semantic-release version.

Thanks for sharing! I might be wrong, but it doesn't look like the code does a major bump on `...!: ...` and `BREAKING CHANGES` style commits?

**Comment by webpro** [comment]
Maybe the function can be extended; I'll be happy to add it to the docs or maybe even merge it into the codebase (with some logic/option to enable it).

**Comment by Blackclaws** [comment]
@webpro This function doesn't work for preRelease and null is ignored there. The . version is still being incremented.

**Comment by Blackclaws** [comment]
This is fixed in #102

**Comment by webpro** [comment]
Looks like we forgot to close this one. If there's still any issues, feel free to open a new ticket.
